tests/autoscaler/autoscaler_test.go

```diff
@@ -0,0 +1,199 @@
+
+	// Get number of worker nodes.
+	workersCount, err := getWorkersCount(ctx, tcCtrlClient)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	logger.Debugf(ctx, "Found %d worker nodes", workersCount)
+
+	// Scale helloworld deployment to len(workers) + 1 replicas to trigger a scale up.
+	expectedWorkersCount := int32(workersCount + 1)
+	logger.Debugf(ctx, "Scaling deployment %s/%s to %d replicas", helloWorldNamespace, helloWorldDeploymentName, expectedWorkersCount)
+	err = scaleDeployment(ctx, tcCtrlClient, expectedWorkersCount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	logger.Debugf(ctx, "Waiting for %d worker nodes to exist", expectedWorkersCount)
+
+	// Wait for nodes to increase by one.
+	o := func() error {
+		workersCount, err := getWorkersCount(ctx, tcCtrlClient)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if int32(workersCount) != expectedWorkersCount {
+			return microerror.Maskf(executionFailedError, "Expecting %d workers, %d found", expectedWorkersCount, workersCount)
+		}
+
+		return nil
+	}
+	b := backoff.NewConstant(backoff.LongMaxWait, 10*time.Second)
+	n := backoff.NewNotifier(logger, ctx)
+	err = backoff.RetryNotify(o, b, n)
+	if err != nil {
+		t.Fatalf("timeout waiting for cluster to scale up: %v", err)
+	}
+
+	// Scale down deployment, wait for one node to get deleted.
+	expectedWorkersCount = expectedWorkersCount - 1
+	logger.Debugf(ctx, "Scaling deployment %s/%s to %d replicas", helloWorldNamespace, helloWorldDeploymentName, expectedWorkersCount)
+	err = scaleDeployment(ctx, tcCtrlClient, expectedWorkersCount)
+	if err != nil {
+		t.Fatalf("timeout waiting for cluster to scale down: %v", err)
+	}
+
+	err = backoff.RetryNotify(o, b, n)
+	if err != nil {
+		t.Fatalf("timeout waiting for cluster to scale down: %v", err)
+	}
+}
+
+func getWorkersCount(ctx context.Context, ctrlClient client.Client) (int, error) {
+	workers := &corev1.NodeList{}
+	err := ctrlClient.List(ctx, workers, client.MatchingLabels{"kubernetes.io/role": "worker"})
+	if err != nil {
+		return -1, err
+	}
+
+	return len(workers.Items), nil
+}
+
+func scaleDeployment(ctx context.Context, ctrlClient client.Client, expectedWorkersCount int32) error {
+	deployment := &appsv1.Deployment{}
+	err := ctrlClient.Get(ctx, client.ObjectKey{Namespace: helloWorldNamespace, Name: helloWorldDeploymentName}, deployment)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	deployment.Spec.Replicas = &expectedWorkersCount
+
+	err = ctrlClient.Update(ctx, deployment)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func createDeployment(ctx context.Context, ctrlClient client.Client, replicas int32) (*appsv1.Deployment, error) {
+	// Check if deployment exists already.
+	existing := &appsv1.Deployment{}
+	err := ctrlClient.Get(ctx, client.ObjectKey{Name: helloWorldDeploymentName, Namespace: helloWorldNamespace}, existing)
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, microerror.Mask(err)
+	}
+
+	if existing.Name != "" {
```

Comment:
Is this `if` required? At this point we already know it was found. When would `existing.Name` be empty?

Reply:
no, the deployment can very well be missing, that's why I check the error to not be "NotFound" in the if above

Reply:
I meant the `if existing.Name != ""` . Why do we check that?

Reply:
to distinguish between when the deployment exists and when it doesn't. Am I missing something obvious?

Reply:
We talked on Slack and we agreed I don't know how to read. Move along.